### PR TITLE
Support reverse reads in the row cache natively

### PR DESCRIPTION
**Pull request by tgrabiec**
This change makes row cache support reverse reads natively so that reversing wrappers are not needed when reading from cache and thus the read can be executed efficiently, with similar cost as the forward-order read. 

The database is serving reverse reads from cache by default after this. Before, it was bypassing cache by default after 703aed327751bcf048f7afd4268497639dab8bca.

Refs: #1413

Tests:

  - unit [dev]
  - manual query with build/dev/scylla and cache tracing on

### CODE REVIEW

**Comment by tgrabiec** [comment]
I don't see any sign of regression, there's even a slight and consistent improvement despite an increase in insns/op. Maybe icache related.

`perf_simple_query -c1 -m1G --duration 60`:

Before:
```
median 115995.41 tps ( 76.1 allocs/op,  12.1 tasks/op,   45539 insns/op)
median absolute deviation: 106.40
maximum: 116246.75
minimum: 114993.09
```

After:
```
median 117330.23 tps ( 76.1 allocs/op,  12.1 tasks/op,   45641 insns/op)
median absolute deviation: 133.44
maximum: 118031.00
minimum: 117004.90
```

**Comment by bhalevy** [comment]
ping...

@tgrabiec et al., what's missing to complete this PR?

**Comment by tgrabiec** [comment]
@bhalevy As far as I know there are no outstanding comments to address. There was a request for this to be reviewed by @haaawk

**Comment by bhalevy** [comment]
@haaawk ping, please review.

**Comment by haaawk** [comment]
Sorry for the delay. Will try to look at it tomorrow.

File: test/boost/row_cache_test.cc [hunk 1]
```diff
@@ -3360,8 +3360,6 @@ SEASTAR_TEST_CASE(test_concurrent_reads_and_eviction) {
         auto pr = dht::partition_range::make_singular(m0.decorated_key());
         auto make_reader = [&] (const query::partition_slice& slice) {
             auto rd = cache.make_reader(s, semaphore.make_permit(), pr, slice);
-            rd.set_max_buffer_size(3);
```
Comment (haaawk) [inline test/boost/row_cache_test.cc:3218]: Should we do the same fix in [test_readers_get_all_data_after_eviction](https://github.com/scylladb/scylla/blob/29fcfe5b1c93406c87c2deeeac8b05933be2cb24/test/boost/row_cache_test.cc#L2278-L2279), [test_tombstones_are_not_missed_when_range_is_invalidated](https://github.com/scylladb/scylla/blob/29fcfe5b1c93406c87c2deeeac8b05933be2cb24/test/boost/row_cache_test.cc#L2407-L2408), [test_exception_safety_of_update_from_memtable](https://github.com/scylladb/scylla/blob/29fcfe5b1c93406c87c2deeeac8b05933be2cb24/test/boost/row_cache_test.cc#L2512-L2513), [test_concurrent_population_before_latest_version_iterator](https://github.com/scylladb/scylla/blob/29fcfe5b1c93406c87c2deeeac8b05933be2cb24/test/boost/row_cache_test.cc#L2725-L2726), [test_random_row_population](https://github.com/scylladb/scylla/blob/29fcfe5b1c93406c87c2deeeac8b05933be2cb24/test/boost/row_cache_test.cc#L2889-L2890), [test_continuity_is_populated_when_read_overlaps_with_older_version?](https://github.com/scylladb/scylla/blob/29fcfe5b1c93406c87c2deeeac8b05933be2cb24/test/boost/row_cache_test.cc#L3015-L3016), [test_continuity_population_with_multicolumn_clustering_key](https://github.com/scylladb/scylla/blob/29fcfe5b1c93406c87c2deeeac8b05933be2cb24/test/boost/row_cache_test.cc#L3147-L3148), [test_tombstone_merging_of_overlapping_tombstones_in_many_versions](https://github.com/scylladb/scylla/blob/29fcfe5b1c93406c87c2deeeac8b05933be2cb24/test/boost/row_cache_test.cc#L3323-L3324) and [test_random_population_with_many_versions](https://github.com/scylladb/scylla/blob/29fcfe5b1c93406c87c2deeeac8b05933be2cb24/test/boost/row_cache_test.cc#L3658-L3659) ?

File: test/boost/mvcc_test.cc [hunk 2]
```diff
@@ -863,6 +863,410 @@ SEASTAR_TEST_CASE(test_partition_snapshot_row_cursor) {
     });
 }
 
+SEASTAR_TEST_CASE(test_partition_snapshot_row_cursor_reversed) {
+    return seastar::async([] {
```
Comment (haaawk) [inline test/boost/mvcc_test.cc:961]: Nit: SEASTAR_THREAD_TEST_CASE?

File: test/boost/mvcc_test.cc [hunk 3]
```diff
@@ -863,6 +863,410 @@ SEASTAR_TEST_CASE(test_partition_snapshot_row_cursor) {
...
+                BOOST_REQUIRE(cur.maybe_refresh());
+                BOOST_REQUIRE(eq(cur.table_position(), table.make_ckey(ck_5)));
+                BOOST_REQUIRE(!cur.continuous());
+            }
+        });
+    });
+}
+
+SEASTAR_TEST_CASE(test_cursor_tracks_continuity_in_reversed_mode) {
+    return seastar::async([] {
```
Comment (haaawk) [inline test/boost/mvcc_test.cc:1152]: NIT: SEASTAR_THREAD_TEST_CASE?

File: test/boost/mvcc_test.cc [hunk 4]
```diff
@@ -863,6 +863,410 @@ SEASTAR_TEST_CASE(test_partition_snapshot_row_cursor) {
...
+                BOOST_REQUIRE(cur.continuous());
+
+                BOOST_REQUIRE(!cur.next());
+            }
+        });
+    });
+}
+
+SEASTAR_TEST_CASE(test_ensure_entry_in_latest_in_reversed_mode) {
+    return seastar::async([] {
```
Comment (haaawk) [inline test/boost/mvcc_test.cc:1259]: NIT: SEASTAR_THREAD_TEST_CASE?

File: test/boost/mvcc_test.cc [hunk 5]
```diff
@@ -863,6 +863,410 @@ SEASTAR_TEST_CASE(test_partition_snapshot_row_cursor) {
...
+                    auto res = cur.ensure_entry_in_latest();
+                    BOOST_REQUIRE(!res.inserted);
+                }
+            }
+        });
+    });
+}
+
+SEASTAR_TEST_CASE(test_ensure_entry_in_latest_does_not_set_continuity_in_reversed_mode) {
+    return seastar::async([] {
```
Comment (haaawk) [inline test/boost/mvcc_test.cc:1314]: NIT: SEASTAR_THREAD_TEST_CASE?

File: partition_snapshot_row_cursor.hh [hunk 6]
```diff
@@ -111,52 +111,94 @@ class partition_snapshot_row_cursor final {
...
+    position_in_partition_view to_table_domain(position_in_partition_view pos) const {
+        if (_reversed) [[unlikely]] {
+            return pos.reversed();
+        }
+        return pos;
+    }
+
+    position_in_partition_view to_query_domain(position_in_partition_view pos) const {
+        if (_reversed) [[unlikely]] {
+            return pos.reversed();
```
[the remark this answers is not in the document]
Comment (haaawk) [reply partition_snapshot_row_cursor.hh:162]: It would be even better to enforce the meaning of input and output using type system.

File: partition_snapshot_row_cursor.hh [hunk 7]
```diff
@@ -428,7 +587,9 @@ public:
     // Assumes the snapshot is evictable and not populated by means other than ensure_entry_if_complete().
     // Subsequent calls to ensure_entry_if_complete() must be given strictly monotonically increasing
     // positions unless iterators are invalidated across the calls.
+    // The cursor must not be a reversed-order cursor.
     std::optional<ensure_result> ensure_entry_if_complete(position_in_partition_view pos) {
+        assert(!_reversed); // latest_i is unreliable
```
Comment (haaawk) [inline partition_snapshot_row_cursor.hh:598]: Wouldn't it be better to use `on_internal_error`?

Comment (haaawk) [reply partition_snapshot_row_cursor.hh:598]: That would crash the query but not the node.

File: partition_snapshot_row_cursor.hh [hunk 8]
```diff
@@ -306,35 +434,44 @@ public:
     // Assumes that rows are not inserted into the snapshot (static). They can be removed.
     bool maybe_refresh_static() {
         if (!iterators_valid()) {
-            return maybe_advance_to(_position);
+            return maybe_advance_to(to_query_domain(_position));
```
Comment (haaawk) [inline partition_snapshot_row_cursor.hh:443]: Would it make sense to replace `to_query_domain(_position)` with `position()`?

File: partition_snapshot_row_cursor.hh [hunk 9]
```diff
@@ -401,21 +538,43 @@ public:
...
             // Copy row from older version because rows in evictable versions must
             // hold values which are independently complete to be consistent on eviction.
-            auto e = current_allocator().construct<rows_entry>(_schema, *_current_row[0].it);
-            e->set_continuous(latest_i && latest_i->continuous());
-            _snp.tracker()->insert(*e);
-            rows.insert_before(latest_i, *e);
-            return {*e, true};
+            auto e = [&] {
+                if (!at_a_row()) {
+                    return alloc_strategy_unique_ptr<rows_entry>(
```
Comment (haaawk) [inline partition_snapshot_row_cursor.hh:559]: What does this alloc_strategy_unique_ptr do and why is it needed in the new version and not needed in the previous version?

File: partition_snapshot_row_cursor.hh [hunk 10]
```diff
@@ -401,21 +538,43 @@ public:
...
         } else {
             // Copy row from older version because rows in evictable versions must
             // hold values which are independently complete to be consistent on eviction.
-            auto e = current_allocator().construct<rows_entry>(_schema, *_current_row[0].it);
-            e->set_continuous(latest_i && latest_i->continuous());
-            _snp.tracker()->insert(*e);
-            rows.insert_before(latest_i, *e);
-            return {*e, true};
+            auto e = [&] {
+                if (!at_a_row()) {
```
Comment (haaawk) [inline partition_snapshot_row_cursor.hh:558]: Why was it not possible before that `_current_row.empty()` and now it is?

File: partition_snapshot_row_cursor.hh [hunk 11]
```diff
@@ -198,17 +268,41 @@ class partition_snapshot_row_cursor final {
...
+                        curr.it = {};
+                    } else {
+                        curr.continuous = curr.it->continuous();
+                        --curr.it;
+                    }
+                } else {
+                    ++curr.it;
+                    if (curr.it) {
+                        curr.continuous = curr.it->continuous();
+                    }
```
Comment (haaawk) [inline partition_snapshot_row_cursor.hh:305]: This was not present in the previous version and it does not seem related to reversed queries. Why does it have to be added?

File: partition_snapshot_row_cursor.hh [hunk 12]
```diff
@@ -198,17 +268,41 @@ class partition_snapshot_row_cursor final {
...
+                }
             }
             if (curr.version_no == 0) {
-                _latest_it = curr.it;
+                if (curr.it) { // in _reversed case above the iterator may be cleared
+                    _latest_it = curr.it;
+                }
             }
```
Comment (haaawk) [inline partition_snapshot_row_cursor.hh:312]: If you push this `if` into the `if (curr.it)` below, you won't need to check `if (curr.it)` twice.

File: partition_snapshot_row_cursor.hh [hunk 13]
```diff
@@ -165,24 +207,52 @@ class partition_snapshot_row_cursor final {
...
         _heap.clear();
         _latest_it.reset();
+        _background_continuity = false;
         int version_no = 0;
         bool unique_owner = _unique_owner;
         bool first = true;
         for (auto&& v : _snp.versions()) {
             unique_owner = unique_owner && (first || !v.is_referenced());
-            auto rows = v.partition().clustered_rows();
+            auto& rows = v.partition().mutable_clustered_rows();
```
[the remark this answers is not in the document]
Comment (haaawk) [reply partition_snapshot_row_cursor.hh:230]: Why does `--` on an iterator require mutable collection?

File: partition_snapshot_row_cursor.hh [hunk 14]
```diff
@@ -111,52 +111,94 @@ class partition_snapshot_row_cursor final {
...
     // Removes the next row from _heap and puts it into _current_row
     bool recreate_current_row() {
         _current_row.clear();
+        _continuous = _background_continuity;
+        _dummy = true;
         if (_heap.empty()) {
+            if (_reversed) {
+                _position = position_in_partition::before_all_clustered_rows();
+            } else {
+                _position = position_in_partition::after_all_clustered_rows();
```
Comment (haaawk) [inline partition_snapshot_row_cursor.hh:191]: This line affects only non-reversed reads. Was it a bug that it was missing previously?

File: partition_snapshot_row_cursor.hh [hunk 15]
```diff
@@ -111,52 +111,94 @@ class partition_snapshot_row_cursor final {
...
-        _dummy = true;
+        version_heap_less_compare heap_less(*this);
+        position_in_partition::equal_compare eq(*_snp.schema());
         do {
             boost::range::pop_heap(_heap, heap_less);
             memory::on_alloc_point();
             rows_entry& e = *_heap.back().it;
             _dummy &= bool(e.dummy());
-            _continuous |= bool(e.continuous());
+            _continuous |= bool(_heap.back().continuous);
```
Comment (haaawk) [inline partition_snapshot_row_cursor.hh:202]: What's the difference between `_heap.back().continuous` and `_heap.back().it->continuous()`?

File: partition_snapshot_row_cursor.hh [hunk 16]
```diff
@@ -641,7 +641,7 @@ public:
     friend std::ostream& operator<<(std::ostream& out, const partition_snapshot_row_cursor& cur) {
         out << "{cursor: position=" << cur._position << ", cont=" << cur.continuous() << ", ";
         if (cur._reversed) {
-            out << "rev=1, ";
+            out << "reversed, ";
```
Comment (haaawk) [inline partition_snapshot_row_cursor.hh:650]: I think this can be just push to the previous patch when it was introduced.

File: partition_version.cc [hunk 17]
```diff
@@ -586,10 +611,9 @@ partition_snapshot::range_tombstones(position_in_partition_view start, position_
     while (!streams.empty()) {
         std::pop_heap(streams.begin(), streams.end(), stream_less);
         range_tombstone_list::iterator_range& stream = streams.back();
-        if (callback(stream.begin()->tombstone()) == stop_iteration::yes) {
+        if (callback(pop_stream(stream)) == stop_iteration::yes) {
             return stop_iteration::no;
         }
-        stream.advance_begin(1);
```
Comment (haaawk) [inline partition_version.cc:592]: This changes the behaviour also when `reversed == false`. Was the previous behaviour incorrect?

Comment (haaawk) [reply partition_version.cc:592]: stream.advance_begin(1) wasn't called when `callback(stream.begin()->tombstone()) == stop_iteration::yes` previously

Comment (haaawk) [reply partition_version.cc:592]: ok. I see that `streams` is a local variable so its state does not matter after the function returns.

File: cache_flat_mutation_reader.hh [hunk 18]
```diff
@@ -383,21 +416,27 @@ future<> cache_flat_mutation_reader::read_from_underlying() {
...
+                                    // It preserves the continuity of the range the entry falls into.
                                     it->set_continuous(next->continuous());
                                     clogger.trace("csm {}: inserted empty row at {}, cont={}", fmt::ptr(this), it->position(), it->continuous());
                                 }
                             });
                         } else if (ensure_population_lower_bound()) {
                             with_allocator(_snp->region().allocator(), [&] {
                                 auto e = alloc_strategy_unique_ptr<rows_entry>(
-                                    current_allocator().construct<rows_entry>(*_schema, _upper_bound, is_dummy::yes, is_continuous::yes));
+                                    current_allocator().construct<rows_entry>(table_s, to_table_domain(_upper_bound), is_dummy::yes, is_continuous::no));
```
Comment (haaawk) [inline cache_flat_mutation_reader.hh:429]: Could you please explain why the entry is no longer continuous?

File: cache_flat_mutation_reader.hh [hunk 19]
```diff
@@ -452,7 +492,13 @@ void cache_flat_mutation_reader::maybe_update_continuity() {
     if (can_populate() && ensure_population_lower_bound()) {
         with_allocator(_snp->region().allocator(), [&] {
             rows_entry& e = _next_row.ensure_entry_in_latest().row;
-            e.set_continuous(true);
+            if (_read_context.is_reversed()) {
```
Comment (haaawk) [inline cache_flat_mutation_reader.hh:496]: [[unlikely]]?

File: cache_flat_mutation_reader.hh [hunk 20]
```diff
@@ -500,8 +546,13 @@ void cache_flat_mutation_reader::maybe_add_to_cache(const clustering_row& cr) {
 
         rows_entry& e = *it;
         if (ensure_population_lower_bound()) {
-            clogger.trace("csm {}: set_continuous({})", fmt::ptr(this), e.position());
-            e.set_continuous(true);
+            if (_read_context.is_reversed()) {
```
Comment (haaawk) [inline cache_flat_mutation_reader.hh:550]: [[unlikely]] ?

File: row_cache.cc [hunk 21]
```diff
@@ -43,9 +43,17 @@ logging::logger clogger("cache");
 using namespace std::chrono_literals;
 using namespace cache;
 
+static schema_ptr to_read_domain(const query::partition_slice& slice, schema_ptr table_domain_schema) {
```
Comment (haaawk) [inline row_cache.cc:46]: Is `read domain` the same thing as `query domain`? If so, why use different names?

**Comment by haaawk** [comment]
LGTM in general but I will need another pass. There's a lot to digest here.

File: partition_snapshot_row_cursor.hh (hunk 6 above)
[the remark this answers is not in the document]
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:162]: Yes, it would.

File: row_cache.cc (hunk 21 above)
Comment (tgrabiec) [reply row_cache.cc:46]: It is the same, I'll switch to "query domain".

File: cache_flat_mutation_reader.hh (hunk 20 above)
Comment (tgrabiec) [reply cache_flat_mutation_reader.hh:550]: Ok

File: cache_flat_mutation_reader.hh (hunk 19 above)
Comment (tgrabiec) [reply cache_flat_mutation_reader.hh:496]: Ok

File: cache_flat_mutation_reader.hh (hunk 18 above)
Comment (tgrabiec) [reply cache_flat_mutation_reader.hh:429]: If we're populating in reverse, we haven't yet read the part of the query range which the is_continuous::yes flag marks (for keys smaller than the entry's key). 

The continuity will be set properly a few lines below, immediately:

```
                               if (_read_context.is_reversed()) {
                                    clogger.trace("csm {}: set_continuous({})", fmt::ptr(this), _last_row.position());
                                    _last_row->set_continuous(true);
                                } else {
                                    clogger.trace("csm {}: mark {} as continuous", fmt::ptr(this), insert_result.first->position());
                                    clogger.trace("csm {}: set_continuous({})", fmt::ptr(this), insert_result.first->position());
                                    insert_result.first->set_continuous(true);
                                }
``` 

The second branch makes the behavior in forward mode equivalent to what was before, so we lose nothing.

File: partition_snapshot_row_cursor.hh (hunk 16 above)
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:650]: right

File: partition_snapshot_row_cursor.hh (hunk 15 above)
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:202]: Both flags determine continuity for the range in the direction of lower key values, the difference is in which schema domain is used for ordering. `_heap.back().continuous` is using cursor schema domain, so in reverse mode, the direction is towards higher key values in table schema domain, whereas `it->continuous()` is using table schema domain (it's as stored in memory, so schema domain is fixed).

File: partition_snapshot_row_cursor.hh (hunk 14 above)
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:191]: There was no bug. This change is in order to support the new, more relaxed, semantics for maybe_advance_to():

>  // If no such entry exists, the cursor is positioned at an extreme position in the direction of
>     // the cursor (min for reversed cursor, max for forward cursor) and not pointing at a row
>     // but still valid.
>    

It now allows advancing even if there is no dummy entry. Forward reads don't rely on it yet, but it's good to be symmetric.

File: partition_snapshot_row_cursor.hh (hunk 13 above)
[the remark this answers is not in the document]
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:230]: Good question, not sure. I will investigate this more when I revisit this.

File: partition_snapshot_row_cursor.hh (hunk 11 above)
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:305]: `curr.continuous` is related to reversed queries. We also discuss this in this thread: https://github.com/scylladb/scylla/pull/9454#discussion_r763213161, check it out.

Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:305]: curr.continuous is a new thing. I added it so that it uniformly reflects cursor's continuity from both forward and reverse perspective. it->continuous() reflects only forward perspective (table order)

File: partition_snapshot_row_cursor.hh (hunk 10 above)
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:558]: advance_to() can now advance to a position which is not a row. We need that because in reverse view there is no dummy before all rows (in table order that is, "after all rows" in read order).

See https://github.com/scylladb/scylla/pull/9454#discussion_r763216731

File: partition_snapshot_row_cursor.hh (hunk 9 above)
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:559]: It's like std::unique_ptr but invokes current_allocator() to delete stuff (so can be used with LSA-managed objects)

I need it because I call a throwing insert() below in the reverse path, and we need to free the object on exception from insert().

Turns out that before we also needed it, because insert_before() is throwing too, @xemul fixed it recently.

File: partition_snapshot_row_cursor.hh (hunk 8 above)
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:443]: Yes

File: partition_snapshot_row_cursor.hh (hunk 7 above)
Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:598]: Yes

File: test/boost/mvcc_test.cc (hunk 5 above)
Comment (tgrabiec) [reply test/boost/mvcc_test.cc:1314]: I'll blanket convert the whole file after this PR, ok?

File: test/boost/row_cache_test.cc (hunk 1 above)
Comment (tgrabiec) [reply test/boost/row_cache_test.cc:3218]: Yes

**Comment by tgrabiec** [comment]
In V2:
  - dropped patch "tests: row_cache_test: Avoid missing close() on the reader if fill_buffer() throws": the fix will be sent separately and will cover more tests
  - Avoided change in partition_snapshot_row_cursor from clustered_rows() to mutable_clustered_rows().
  - Introduced "utils: immutable-collection: Make movable" (needed by the above)
  - Squashed the hunk which changes logging from "rev=1" to "reversed"
  - Consistently use the name "to_query_domain" instead of "to_read_domain"
  - Sprinkle more [[unlikely]] to reversed branch
  - Use position() call instead of open coding conversion to query domain in partition_snapshot_row_cursor
  - Moved `if (curr.version_no)` under `if (curr.it) ` in partition_snapshot_row_cursor::advance()
  - Throw instead of assert in ensure_entry_if_complete()

File: db/config.cc [hunk 22]
```diff
@@ -810,7 +810,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
...
     , reversed_reads_auto_bypass_cache(this, "reversed_reads_auto_bypass_cache", liveness::LiveUpdate, value_status::Used, true,
-            "Use a new implementation of reversed reads in sstables when performing reversed queries. The new implementation does not require unbounded memory (compared to the old implementation which had to fetch entire partitions into memory) but disables the cache. Turn this option off if your partitions are small so the old implementation is good enough; your queries can then utilize the cache and potentially be faster (since they don't need to use sstables as much). This option is temporary and will be removed as soon as the cache and sstable reverse read algorithms are updated to handle reversed queries correctly.")
+            "Bypass in-memory data cache (the row cache) when performing reversed queries.")
```
[the remark this answers is not in the document]
Comment (kbr-) [reply db/config.cc:842]: Right now the `reversed_reads_auto_bypass_cache` option will use the old reversing reader when `false` and perform a regular forward query underneath. This means that the sstable query will be done in forward mode. So this option also gives us the possibility to disable the sstable reverse reading code in case we find that there are bugs in it.

I think we should keep an option for falling back to the old reversing reader for a while.

Anyway, at this point the option is called `reversed_reads_auto_bypass_cache` and the default is `true`. Which would mean that we automatically bypass the cache during reversed queries by default even with this PR?

[the remark this answers is not in the document]
Comment (tgrabiec) [reply db/config.cc:842]: > Right now the `reversed_reads_auto_bypass_cache` option will use the old reversing reader when `false` and perform a regular forward query underneath. This means that the sstable query will be done in forward mode. So this option also gives us the possibility to disable the sstable reverse reading code in case we find that there are bugs in it.
> 
> I think we should keep an option for falling back to the old reversing reader for a while.

I agree that it would be good to have an option for disabling reverse readers globally, which would also avoid memtable reverse readers, not just sstable reverse readers. 

> 
> Anyway, at this point the option is called `reversed_reads_auto_bypass_cache` and the default is `true`. Which would mean that we automatically bypass the cache during reversed queries by default even with this PR?

The default is changed to false in https://github.com/tgrabiec/scylla/commit/65a1a0247a689eb44c0986b7a3cc65fc356a8798

File: partition_snapshot_row_cursor.hh [hunk 23]
```diff
@@ -111,52 +111,94 @@ class partition_snapshot_row_cursor final {
     friend class partition_snapshot_row_weakref;
     struct position_in_version {
         mutation_partition::rows_type::iterator it;
+        utils::immutable_collection<mutation_partition::rows_type> rows;
```
Comment (avikivity) [inline partition_snapshot_row_cursor.hh:120]: It's odd that position_in_version (an iterator-like thing) has a rows container member. What's the explanation? 

This deserves a comment.

Comment (avikivity) [reply partition_snapshot_row_cursor.hh:120]: It's odd because immutable_collection should be named immutable_view_of_collection. That of course can be done later.

Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:120]: > It's odd that position_in_version (an iterator-like thing) has a rows container member. What's the explanation?
> 

I need to compare against rows.begin() during backwards advance(). In forward mode this is not needed since d891cfe6cdd9215613a0ebeeaeb158f4bbc4be5d due to the fact that btree iterators become empty when they reach end()

Comment (tgrabiec) [reply partition_snapshot_row_cursor.hh:120]: > It's odd because immutable_collection should be named immutable_view_of_collection. That of course can be done later.

Yes, it's a preexisting problem.
